src/NHibernate.Test/NHSpecificTest/GH3539/CardInfo.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NHibernate.Test.NHSpecificTest.GH3539;
+
+public class CardInfo
+{
+	private readonly ISet<string> _oldCards;
+
+	protected CardInfo() { }
+
+	public CardInfo(params string[] cards)
+	{
+		_oldCards = cards.ToHashSet();
+	}
+
+	public virtual ISet<string> GetCardsCopy()
+	{
+		return _oldCards.ToHashSet();
+	}
+
+	public override bool Equals(object obj)
+	{
+		if (obj is null)
+		{
+			return false;
+		}
+		if (ReferenceEquals(this, obj))
+		{
+			return true;
+		}
+		if (obj.GetType() != GetType())
+		{
+			return false;
+		}
+		var other = (CardInfo) obj;
+		return _oldCards.SetEquals(other._oldCards);
+	}
+
+	public override int GetHashCode()
+	{
+		var hashCode = new HashCode();
+		hashCode.Add(_oldCards);
```

Comment:
You have invalid hashcode generation function. You add a collection as an object.
```suggestion
		foreach (var card in _oldCards) hashCode.Add(card);
```

Reply:
Thank you, I agree. But this is not a production code of course. I don't think it changes anything regarding the issue. Would just be a performance issue on production, hash collision.

Reply:
NHibernate uses GetHashCode to determine equality. It first checks hash code (must be the same), then calls equals.

Reply:
oh, right! It actually may produce different hash codes for what we think should be the same value. I think the best approach would have been to use HashSet<String>.CreateSetComparer().GetHashCode(_oldCards). But it is not possible because NHibernate requires _oldCards to be an ISet. It uses foreach underneath anyway.